### PR TITLE
Refactor .travis.yml

### PR DESCRIPTION
**Pull request by peter-edge**
@abhinav @breerly I was going to type up a long description to explain this, but it would be way way easier to talk through this, ping me when you get in if you can.

Note on my local machine, `make ci` takes about 2m30s from scratch, and about 1m40s after running once.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,128 +2,45 @@ sudo: false
...
         - secure: dhvshW7GvphJ6xvS8MxEtJGN+4C4d64hqOcaCKXXXVqyeunJ1gH/Cb6v1RcETgtRxOIHyyHo218j0cs0hNtIWCZq5YVbXMAJ5Sret+cKA3aZNMCbgwPNS48WdVZsCHurW22abq41ejfKuuFqSI+ooMprrsGeRjHMbJZM86k7UuEqxdvqGjotYugGcfiCjxnNtMHd1M6AQ9fPFzhg8TUeqPy7eDIdKTcqwa/h7ZsYn/WJp1LkUtwyHsfpweIv6JBQ9/xUDzjcoz5MA6Nj75/Mk4bJzzLEem3EKs1TTZOqV1NVt7VJ887NPe4fLIxP1/jQLzsh6kShTcctLszc70NgrowdwhdIqN/3TUmxkVbGP2JzbQ4yR6LVKCDDeWpKGiPsIKy75kEiW4ZlEZhnwCOKaR94FyBy1HEnhyoD9OJD6Gq/u4UdEw/QuNfE89gbr47LA4z487AzYGzH+1J3MlEbTNgTCmWduC0ghTx33vItrYMXXabGDe7XwfdJKTnMjfXVv+CVBdMAZVtDL27sEmbbAzIrRjP+JOFm7ZRYfzvxu5vCWlB/3wMMt7DzcutjxR4tyYs71Ec5zdUH/16wSvjQR1laJZCujlNSA/3FZUizCzW3qFVsBkaWRF+/C/cksi6cR83xIrc4y4WRLttAcbfqmsXaigL5Iz4GbLg62vmFP88=
-    - go: 1.8
-      services: [docker]
-      sudo: required
+      cache:
+        directories:
+          - .cache
```
Comment (breerly) [inline .travis.yml]: nice

Comment (breerly) [inline .travis.yml]: I think there is a way to get docker images to cache so that we don't pull down the same crossdock images every time.

**Comment by peter-edge** [comment]
Ya let's catch up offline - summary is this actually fixes the build,
builds haven't been running on 1.8 for a while, not kidding, this fixes that

File: Makefile [hunk 2]
```diff
@@ -171,19 +234,40 @@ docker-build:
...
-	else \
-		echo "Version number in version.go does not match CHANGELOG.md"; \
-		echo "version.go: $(INTHECODE_VERSION)"; \
-		echo "CHANGELOG : $(CHANGELOG_VERSION)"; \
-		exit 1; \
-	fi
+.PHONY: ci-docker-load
+ci-docker-load:
+ifdef CI_CROSSDOCK
+	if [ -f $(CI_DOCKER_CACHE_FILE) ]; then gunzip -c $(CI_DOCKER_CACHE_FILE) | docker load; fi
```
Comment (abhinav) [inline Makefile]: cool

File: .travis.yml [hunk 3]
```diff
@@ -2,128 +2,45 @@ sudo: false
...
-        if [ -n "$DOCKER_EMAIL" ] && [ -n "$DOCKER_USER" ] && [ -n "$DOCKER_PASS" ]; then
-          docker login -e $DOCKER_EMAIL -u $DOCKER_USER -p $DOCKER_PASS
-          docker push $REPO
-        fi
-      fi
-      ;;
-  esac
-  set +e
+  - make ci
+  - if [[ "${TRAVIS_BRANCH}" == "master" ]] && [[ "${TRAVIS_PULL_REQUEST}" == "false" ]]; then make ci-docker-save; fi
```
Comment (abhinav) [inline .travis.yml]: This isn't right. We want to push docker images for non-master branches too. The idea is that if you're developing a feature across two languages, you can consume the docker image for the feature branch from Go in the feature branch in Java.

**Comment by peter-edge** [comment]
Ok, will re-add that logic, my bad

File: Makefile [hunk 4]
```diff
@@ -41,6 +42,42 @@ FILTER_ERRCHECK := grep -v $(patsubst %,-e %, $(ERRCHECK_EXCLUDES))
...
+ifneq ($(filter goveralls,$(CI_TYPES)),)
+CI_GOVERALLS := true
+endif
+
+CI_CACHE_DIR := $(shell pwd)/.cache
+CI_DOCKER_CACHE_DIR := $(CI_CACHE_DIR)/docker
+CI_DOCKER_IMAGE := yarpc_go
+CI_DOCKER_CACHE_FILE := $(CI_DOCKER_CACHE_DIR)/$(CI_DOCKER_IMAGE)
+
+DOCKER_IMAGE := yarpc/yarpc-go:latest
```
Comment (akshayjshah) [inline Makefile]: Should we use `$(DOCKER_IMAGE)` here too?

File: Makefile [hunk 5]
```diff
@@ -41,6 +42,42 @@ FILTER_ERRCHECK := grep -v $(patsubst %,-e %, $(ERRCHECK_EXCLUDES))
...
+
+DOCKER_COMPOSE_VERSION := 1.10.0
+
+_BIN_DIR = $(CI_CACHE_DIR)/bin
+
+$(_BIN_DIR)/docker-compose:
+	mkdir -p $(_BIN_DIR)
+	curl -L https://github.com/docker/compose/releases/download/$(DOCKER_COMPOSE_VERSION)/docker-compose-$(shell uname -s)-$(shell uname -m) > $(_BIN_DIR)/docker-compose
+	chmod +x $(_BIN_DIR)/docker-compose
```
Comment (akshayjshah) [inline Makefile]: Do we still need this? The Travis Docker environment comes with docker-compose; are we using features that are only available in this particular release?

File: Makefile [hunk 6]
```diff
@@ -98,28 +135,44 @@ govet:
...
+		echo "version.go: $(INTHECODE_VERSION)"; \
+		echo "CHANGELOG : $(CHANGELOG_VERSION)"; \
+		exit 1; \
+	fi
+
+.PHONY: lintbins
+lintbins:
+	@go get github.com/golang/lint/golint
+	@go get honnef.co/go/tools/cmd/staticcheck
+	@go get github.com/kisielk/errcheck
```
Comment (akshayjshah) [inline Makefile]: If you put all the packages in a single `go get` invocation, it's a bit faster (though possibly not for these specific deps) - it lets the go tool compute the whole dependency graph and go get everything just once.

File: Makefile [hunk 7]
```diff
@@ -130,38 +183,48 @@ install:
...
-test: verify_version $(THRIFTRW)
+test: $(THRIFTRW)
 	PATH=$(_GENERATE_DEPS_DIR):$$PATH go test -race $(PACKAGES)
 
 
+.PHONY: coverbins
+coverbins:
+	@go get github.com/wadey/gocovmerge
+	@go get github.com/mattn/goveralls
+	@go get golang.org/x/tools/cmd/cover
```
Comment (akshayjshah) [inline Makefile]: Same here.

File: .travis.yml (hunk 1 above)
Comment (akshayjshah) [inline .travis.yml]: Can we also cache our `gopath/src` and use `go get -u` below? May speed up installing deps that rarely change.

File: Makefile [hunk 8]
```diff
@@ -130,38 +183,48 @@ install:
...
+
 .PHONY: cover
-cover:
-	./scripts/cover.sh $(shell go list $(PACKAGES))
+cover: coverbins $(THRIFTRW)
+	PATH=$(_GENERATE_DEPS_DIR):$$PATH ./scripts/cover.sh $(shell go list $(PACKAGES))
```
Comment (akshayjshah) [inline Makefile]: Use `$PACKAGES` here instead of shelling out to `go list`?

Comment (peter-edge) [inline Makefile]: This was the existing code, can I fix this later?

File: Makefile (hunk 5 above)
Comment (peter-edge) [inline Makefile]: I don't know if we do, but travis was doing it, and this PR is meant for consistency, ie I want to completely guarantee we use the same version of docker-compose locally as we do remotely.

File: Makefile (hunk 6 above)
Comment (peter-edge) [inline Makefile]: Done

File: Makefile (hunk 4 above)
Comment (peter-edge) [inline Makefile]: No, this was for the docker tagging, but doesn't matter anymore.

File: Makefile (hunk 7 above)
Comment (peter-edge) [inline Makefile]: Done

**Comment by peter-edge** [comment]
@akshayjshah we could do the caching potentially, but can we do this later? It's potentially dicey (you're assuming things about travis' environment, specifically one dir in $GOPATH), and this cleanup does enough, I don't want to fix everything if that makes sense, this is already making progress.

The main thing is the next step will be to dockerize everything, so caching `go get` calls will be irrelevant.

File: .travis.yml (hunk 1 above)
Comment (peter-edge) [inline .travis.yml]: See comment below

**Comment by peter-edge** [comment]
All comments should be addressed @abhinav @breerly @akshayjshah

File: .travis.yml [hunk 9]
```diff
@@ -2,128 +2,46 @@ sudo: false
...
-
 matrix:
-  # Fail the build as soon as the first job fails
   fast_finish: true
-
   include:
-    # Crossdock
-    - go: 1.7
-      services: [docker]
+    - services:
```
Comment (abhinav) [inline .travis.yml]: What version of Go will this use? Can we add that explicitly?

File: .travis.yml [hunk 10]
```diff
@@ -43,4 +43,5 @@ install:
   - make install
 script:
   - make ci
+  - scripts/travis-docker-push.sh
```
Comment (breerly) [inline .travis.yml]: This will run, even if the previous line fails my dude.

File: .travis.yml [hunk 11]
```diff
@@ -2,128 +2,46 @@ sudo: false
...
-    # Lint
+        - CI_TYPES='cover goveralls'
     - go: 1.8
-      env: TYPE=lint
-
-    # Tests
-    #
-    # Need virtualenv for example tests
+      env:
+        - CI_TYPES='lint test'
```
Comment (abhinav) [inline .travis.yml]: Can we test with both, 1.7 and 1.8?

**Comment by peter-edge** [comment]
Really? Travis, how you hurt my brain.

File: .travis.yml [hunk 12]
```diff
@@ -2,128 +2,46 @@ sudo: false
...
-      sudo rm -f /usr/local/bin/docker-compose
-      curl -L https://github.com/docker/compose/releases/download/${DOCKER_COMPOSE_VERSION}/docker-compose-`uname -s`-`uname -m` > docker-compose
-      chmod +x docker-compose
-      sudo mv docker-compose /usr/local/bin
-      docker-compose version
-      ;;
-  esac
-  set +e
-
+  - make ci-docker-load
```
Comment (abhinav) [inline .travis.yml]: IMO we should continue mentioning the versions of the tools being used in the Travis logs. Helps debug when there are failures.

**Comment by peter-edge** [comment]
@breerly fixed, note that the docker save call will have to go in the script part because of how travis caching works (I had this issue in FX).

File: Makefile (hunk 5 above)
Comment (akshayjshah) [inline Makefile]: Ah, makes sense. I forgot about the local dev part.

File: Makefile (hunk 8 above)
Comment (akshayjshah) [inline Makefile]: Yep, definitely - landing now sounds reasonable.

**Pull request merged**